datacube/index/_datasets.py

```diff
@@ -703,3 +703,216 @@ def search_eager(self, **query):
+
+        Only the requested fields will be returned together with related derived attributes as property functions
+        similer to the datacube.model.Dataset class. For example, if 'extent'is requested all of
+        'crs', 'extent', 'transform', and 'bounds' are available as property functions.
+
+        The field_names can be custom fields in addition to those specified in metadata_type, fixed fields, or
+        native fields. The field_names can also be derived fields like 'extent', 'crs', 'transform',
+        and 'bounds'. The custom fields require custom offsets of the metadata doc be provided.
+
+        The datasets can be selected based on values of custom fields as long as relevant custom
+        offsets are provided. However custom field values are not transformed so must match what is
+        stored in the database.
+
+        :param field_names: A tuple of field names that would be returned including derived fields
+                            such as extent, crs
+        :param custom_offsets: A dictionary of offsets in the metadata doc for custom fields
+        :param limit: Number of datasets returned per product.
+        :param query: key, value mappings of query that will be processed against metadata_types,
+                      product definitions and/or dataset table.
+        :return: A Dynamically generated DatasetLight (a subclass of namedtuple and possibly with
+        property functions).
+        """
+
+        assert field_names
+
+        class DatasetSpatialMixin(object):
+            __slots__ = ()
+
+            @property
+            def _gs(self):
+                return self.grid_spatial
+
+            @property
+            def crs(self):
+                return Dataset.crs.__get__(self)
+
+            @cached_property
+            def extent(self):
+                return Dataset.extent.__get__(self, Dataset)
+
+            @property
+            def transform(self):
+                return Dataset.transform.__get__(self)
+
+            @property
+            def bounds(self):
+                return Dataset.bounds.__get__(self)
+
+        for product, query_exprs in self.make_query_expr(query, custom_offsets):
+
+            select_fields = self.make_select_fields(product, field_names, custom_offsets)
+            select_field_names = tuple(field.name for field in select_fields)
+            result_type = namedtuple('DatasetLight', select_field_names)
+
+            if 'grid_spatial' in select_field_names:
+                class DatasetLight(result_type, DatasetSpatialMixin):
+                    pass
+            else:
+                class DatasetLight(result_type):
+                    __slots__ = ()
+
+            with self._db.connect() as connection:
+                results = connection.search_unique_datasets(
+                    query_exprs,
+                    select_fields=select_fields,
+                    limit=limit
+                )
+
+            import json
```

Comment:
imports in odd location, move to top of function or module

Reply:
Done